### ISSUE TITLE
Installation Hangs after "Successfully installed Python 2.7" and no log file for visual studio build tools

### ISSUE THREAD

**Issue by jashang87** [felixrieseberg/windows-build-tools#172]
I run the "npm install --global --production windows-build-tool" and installation got stuck after "Successfully installed Python 2.7" (almost 2 hours)
![image](https://user-images.githubusercontent.com/47150101/51918737-154b8d80-23b0-11e9-8f6b-3d24fa003b22.png)
And no log file for studio build tools in ".windows-build-tools" folder
![image](https://user-images.githubusercontent.com/47150101/51918828-3dd38780-23b0-11e9-8d34-7312f40211f7.png)
It installed some programs related to Visual Studio (control panel screenshot below)
![image](https://user-images.githubusercontent.com/47150101/51919029-a02c8800-23b0-11e9-96e0-360373c27339.png)
Any idea how to fix this? Thanks

**Comment by terebentina** [comment]
I am experiencing the same thing. 
Windows 10
nodejs 10.11.0

**Comment by jashang87** [comment]
Hi @terebentina , I figured that everything is installed as required (Python and VS 2017) but the CMD script/installer got stuck and can't get the feedback from VS installer.
So after installer installed the VS 2017 components (as shown in screenshot in my initial post, will take some time to install everything), I opened the task manager and killed "node.exe" processes (two of those processes were running) and then I closed the CMD window, opened a new CMD window and went to next step (in my case, installed node-gyp) and everything worked fine on second attempt (it didn't work first time, I uninstalled everything and followed the process again as I mentioned above).

**Comment by aifodu** [comment]
https://github.com/felixrieseberg/windows-build-tools/issues/152#issuecomment-432945646

**Comment by gian1200** [comment]
I faced a similar issue. As a workaround I installed the 4.0.0 version `npm install --global --production windows-build-tools@4.0.0` and then run the original one `npm install --global --production windows-build-tool`, and it completed

**Comment by deepakjha14** [comment]
I am stuck with the same issue, cant get over "Python 2.7.10 already installed, not installing again" and stays there for the next ice age like a zen monk :( any help will be appreciated, dont know what to do.

**Comment by harmandeep-singh** [comment]
> I faced a similar issue. As a workaround I installed the 4.0.0 version `npm install --global --production windows-build-tools@4.0.0` and then run the original one `npm install -g --production windows-build-tools`, and it completed

Thank you :) it's working

**Comment by peraltafederico** [comment]
Hi everyone! I found the solution to this problem with this steps:

1.  Uninstall Python 2.7
2.  Install [this](https://www.python.org/downloads/release/python-2717/) version of Python
3.  Run `npm install --global windows-build-tools`

Let me know if it works!

**Comment by jonab12** [comment]
@peraltafederico Thank you this worked for me!

**Comment by bkelley13** [comment]
I thought it was stuck too, but it just took about 4 minutes to get past that.

**Comment by ryanbottema** [comment]
> Hi everyone! I found the solution to this problem with this steps:
>
>   Uninstall Python 2.7
>  Install this version of Python
>    Run npm install --global windows-build-tools

This works, but a note to Python3 users; make sure your 'python' path to python2 takes precedence over python3.

**Comment by cSarcasme** [comment]
ty @peraltafederico for your solution stuck now for one hour and you find the solution thank s a lot

**Comment by nowayhaze** [comment]
I'm not sure if I have a legit solution. After waiting for 30 minutes and nothing happening, I just killed the application. And ran the install script from the beginning again. This time it said python was already installed and then blew through the rest of installation.

**Comment by mcaljamj** [comment]
The 5.2.2 package is adding an entry to the users PATH environment variable every time it runs. As was in a CI scenario it was added 10's of time.  Once i removed all unnecessary duplicates the package would again install.

**Comment by Delay-n-days** [comment]
> 我面临着一个类似的问题。作为一个解决方法，我安装了4.0.0版本，然后运行原来的一个，它完成了`npm install --global --production windows-build-tools@4.0.0``npm install -g --production windows-build-tools`

Thanks

**Comment by karlhorky** [comment]
Another option is to try installing the build tools via [Chocolatey](https://chocolatey.org/) instead (use an Administrator command line):

```sh
choco install python visualstudio2017-workload-vctools -y
npm config set msvs_version 2017
```

Maybe this should be documented in the readme as an alternative...

**Comment by Crumor** [comment]
> Another option is to try installing the build tools via [Chocolatey](https://chocolatey.org/) instead (use an Administrator command line):
> 
> ```shell
> choco install python visualcpp-build-tools -y
> npm config set msvs_version 2017
> ```
> 
> Maybe this should be documented in the readme as an alternative...

awesome @karlhorky works like charm thankyou very much

**Comment by FernandoMontesinos** [comment]
> Another option is to try installing the build tools via [Chocolatey](https://chocolatey.org/) instead (use an Administrator command line):
> 
> ```shell
> choco install python visualcpp-build-tools -y
> npm config set msvs_version 2017
> ```
> 
> Maybe this should be documented in the readme as an alternative...

It worked for me, thank you very much

**Comment by Harsh-Khurana** [comment]
Should I delete .windows-built-tools folder and uninstall python 2.7 before re-running the commands that are specified as solutions?

**Comment by j1mmy-Z** [comment]
> Another option is to try installing the build tools via [Chocolatey](https://chocolatey.org/) instead (use an Administrator command line):
> 
> ```shell
> choco install python visualcpp-build-tools -y
> npm config set msvs_version 2017
> ```
> 
> Maybe this should be documented in the readme as an alternative...

It really works, cost me several hours , thanks so much.

**Comment by bodrovdev** [comment]
> Another option is to try installing the build tools via [Chocolatey](https://chocolatey.org/) instead (use an Administrator command line):
> 
> ```shell
> choco install python visualcpp-build-tools -y
> npm config set msvs_version 2017
> ```
> 
> Maybe this should be documented in the readme as an alternative...

THANK YOU!

**Comment by fabiovaz** [comment]
npm config set msvs_version 2017
saved me

**Comment by djbreen7** [comment]
Linking here: https://github.com/felixrieseberg/windows-build-tools/issues/208#issuecomment-877882007

**Comment by bhargu-27** [comment]
> Hi everyone! I found the solution to this problem with this steps:
> 
> 1. Uninstall Python 2.7
> 2. Install [this](https://www.python.org/downloads/release/python-2717/) version of Python
> 3. Run `npm install --global windows-build-tools`
> 
> Let me know if it works!

Hello after installing 2.7.17 when I follow step 3 it again installs 2.7.15

**Comment by jokereey** [comment]
> Another option is to try installing the build tools via [Chocolatey](https://chocolatey.org/) instead (use an Administrator command line):
> 
> ```shell
> choco install python visualstudio2017-workload-vctools -y
> npm config set msvs_version 2017
> ```
> 
> Maybe this should be documented in the readme as an alternative...

It works! thank you man

**Comment by amzbiz** [comment]
Still hangs, doesn't work for me. Off to Linux.

**Comment by karlhorky** [comment]
@amzbiz my [solution above](https://github.com/felixrieseberg/windows-build-tools/issues/172#issuecomment-824916268) is an alternative, not a way to make `windows-build-tools` work.

Once you run those two commands, it has installed the dependencies that `windows-build-tools` is supposed to install.

**Comment by amzbiz** [comment]
Yup I tried with both build tools 2017 and 2019 @karlhorky, one installed directly, and one using your Chocolatey recommendation. Unfortunately neither of those fixed my node 12 build issue. Fortunately I was able to replicate and fix my issue on node 10 but still looks likely I'll need to move to WSL as I don't want to keep pushing things uphill, spent a whole day trying different fixes and none of them got node 12 working 😢.

**Comment by karlhorky** [comment]
Interesting, wonder if the tools are installed properly but Node cannot find them for some reason (a path issue or something). Anyway, good luck!

**Comment by taivasverkko** [comment]
I was struggling with this for awhile but got it working.

First I did the chocolatey recommendation and after that I got error `gyp ERR! UNCAUGHT EXCEPTION`

Workaround for that I found from [here](https://github.com/nodejs/node-gyp/issues/1753#issuecomment-496024604 ).
Run command `npm config set msbuild_path "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin\MSBuild.exe"
`
